hypothesis-python/src/hypothesis/strategies/_internal/utils.py

```diff
@@ -177,7 +178,11 @@ def to_jsonable(obj: object) -> object:
         and dcs.is_dataclass(obj)
         and not isinstance(obj, type)
     ):
-        return to_jsonable(dcs.asdict(obj))
+        if sys.version_info[:2] < (3, 12):
+            # see issue #3812
+            return to_jsonable(dataclass_asdict(obj))
+        else:  # pragma: no cover
+            return to_jsonable(dcs.asdict(obj))
```

Comment:
Let's put this conditional in `compat.py` - we can add 

```python
if sys.version_info[:2] >= (3, 12):
    # see issue #3812
    dataclass_asdict = dataclasses.asdict
```
after the vendored copy.

Reply:
good point.